### PR TITLE
Add RegistrationDeprecationValidator to verify deprecation changes are properly consumed

### PR DESCRIPTION
**Pull request by scottbommarito**
https://github.com/nuget/nugetgallery/issues/7153

### CODE REVIEW

File: src/Catalog/Helpers/FeedHelpers.cs [hunk 1]
```diff
@@ -23,101 +19,5 @@ public static HttpClient CreateHttpClient(Func<HttpMessageHandler> handlerFunc)
...
-
-        /// <summary>
-        /// Asynchronously gets a <see cref="IList{FeedPackageDetails}"/> from the feed.
-        /// </summary>
-        /// <param name="client">An HTTP client.</param>
-        /// <param name="uri">The feed URI.</param>
-        /// <returns>A task that represents the asynchronous operation.
-        /// The task result (<see cref="Task{TResult}.Result" />) returns an
-        /// <see cref="IList{FeedPackageDetails}" />.</returns>
-        public static async Task<IList<FeedPackageDetails>> GetPackages(HttpClient client, Uri uri)
```
Comment (scottbommarito) [inline src/Catalog/Helpers/FeedHelpers.cs]: I believe this was also only used by `Package2CatalogJob`

File: src/Catalog/Helpers/JsonSort.cs [hunk 2]
```diff
@@ -4,8 +4,6 @@
 using System;
 using System.Collections.Generic;
 using System.Linq;
-using System.Text;
```
Comment (scottbommarito) [inline src/Catalog/Helpers/JsonSort.cs]: I did a "remove all unnecessary references" on the solution because I was noticing a lot of them--will gladly revert this change if there's any push-back.

File: src/NuGet.Services.Metadata.Catalog.Monitoring/Model/PackageRegistrationIndexMetadata.cs [hunk 3]
```diff
@@ -14,31 +14,17 @@ namespace NuGet.Services.Metadata.Catalog.Monitoring
     /// </summary>
     public class PackageRegistrationIndexMetadata : PackageRegistrationLeafMetadata
     {
-        public string Authors { get; set; }
```
Comment (scottbommarito) [inline src/NuGet.Services.Metadata.Catalog.Monitoring/Model/PackageRegistrationIndexMetadata.cs]: We used to grab basically every possible field from the V3 registration index, even though we were only using a small subset. Now that the monitoring job is validating against the database and not the feed, in order to continue validating against these fields, we would need to add them to the query used by Db2Catalog. If we decide we would like to add some validations against these fields, we can add them to the query and add them back here.

File: src/NuGet.Services.Metadata.Catalog.Monitoring/Resources/PackageRegistrationMetadataResourceDatabaseFeed.cs [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using NuGet.Common;
+using NuGet.Packaging.Core;
+using NuGet.Services.Metadata.Catalog.Helpers;
+
+namespace NuGet.Services.Metadata.Catalog.Monitoring
+{
+    public class PackageRegistrationMetadataResourceDatabase : IPackageRegistrationMetadataResource
```
Comment (scottbommarito) [inline src/NuGet.Services.Metadata.Catalog.Monitoring/Resources/PackageRegistrationMetadataResourceDatabaseFeed.cs]: This class was renamed from `PackageRegistrationMetadataResourceV2Feed` and had its code changed to pull data from `IGalleryDatabaseQueryService` instead of `V2FeedParser`.

File: src/NuGet.Services.Metadata.Catalog.Monitoring/Validation/Test/Registration/RegistrationDeprecationValidator.cs [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+            if (database.Deprecation.Message != v3.Deprecation.Message)
+            {
+                AddDeprecationInconsistencyException(
+                    exceptions,
+                    database, v3,
+                    nameof(PackageRegistrationDeprecationMetadata.Message),
+                    d => d.Message);
+            }
+
+            if (database.Deprecation.AlternatePackage == null || v3.Deprecation.AlternatePackage == null)
```
Comment (xavierdecoster) [inline src/NuGet.Services.Metadata.Catalog.Monitoring/Validation/Test/Registration/RegistrationDeprecationValidator.cs]: Similar readability issue here. Although technically correct, I think it could be made more readable by simplifying the logical constructs.

```
if (database.Deprecation.AlternatePackage != null && v3.Deprecation.AlternatePackage != null)
{
  if (database.Deprecation.AlternatePackage.Id != v3.Deprecation.AlternatePackage.Id)
  {
    AddAlternatePackageInconsistencyException(
                        exceptions,
                        database, v3,
                        nameof(PackageRegistrationAlternatePackageMetadata.Id),
                        a => a.Id);
  }

  if (database.Deprecation.AlternatePackage.Range != v3.Deprecation.AlternatePackage.Range)
  {
    AddAlternatePackageInconsistencyException(
                        exceptions,
                        database, v3,
                        nameof(PackageRegistrationAlternatePackageMetadata.Range),
                        a => a.Range);
  }
}
else if ((database.Deprecation.AlternatePackage == null && v3.Deprecation.AlternatePackage != null)
  || (database.Deprecation.AlternatePackage != null && v3.Deprecation.AlternatePackage == null))
{
  AddDeprecationInconsistencyException(
                        exceptions,
                        database, v3,
                        nameof(PackageRegistrationDeprecationMetadata.AlternatePackage),
                        d => d.AlternatePackage);
}
```

File: src/NuGet.Services.Metadata.Catalog.Monitoring/Utility/ContainerBuilderExtensions.cs [hunk 6]
```diff
@@ -177,8 +178,8 @@ private static void RegisterDefaultResourceProviders(this ContainerBuilder build
         private static void RegisterV2ResourceProviders(this ContainerBuilder builder)
         {
             builder.RegisterResourceProvider<NonhijackableV2HttpHandlerResourceProvider>(FeedType.HttpV2);
-            builder.RegisterResourceProvider<PackageTimestampMetadataResourceV2Provider>(FeedType.HttpV2);
-            builder.RegisterResourceProvider<PackageRegistrationMetadataResourceV2FeedProvider>(FeedType.HttpV2);
+            builder.RegisterResourceProvider<PackageTimestampMetadataResourceDatabaseProvider>(FeedType.HttpV2);
+            builder.RegisterResourceProvider<PackageRegistrationMetadataResourceDatabaseFeedProvider>(FeedType.HttpV2);
```
[the remark this answers is not in the document]
Comment (scottbommarito) [reply src/NuGet.Services.Metadata.Catalog.Monitoring/Utility/ContainerBuilderExtensions.cs]: good call

File: src/NuGet.Services.Metadata.Catalog.Monitoring/Validation/Test/Exceptions/MetadataFieldInconsistencyException.cs [hunk 7]
```diff
@@ -1,22 +1,23 @@
 ﻿// Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+using Newtonsoft.Json;
 using System;
```
[the remark this answers is not in the document]
Comment (scottbommarito) [reply src/NuGet.Services.Metadata.Catalog.Monitoring/Validation/Test/Exceptions/MetadataFieldInconsistencyException.cs]: VS did indeed reset this option...I hate it

File: src/NuGet.Services.Metadata.Catalog.Monitoring/Validation/Test/Registration/RegistrationDeprecationValidator.cs [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+        {
+        }
+
+        public override Task CompareIndexAsync(ValidationContext context, PackageRegistrationIndexMetadata database, PackageRegistrationIndexMetadata v3)
+        {
+            var exceptions = new List<MetadataFieldInconsistencyException<PackageRegistrationIndexMetadata>>();
+
+            if (database.Deprecation == null || v3.Deprecation == null)
+            {
+                if (database.Deprecation != null || v3.Deprecation != null)
```
[the remark this answers is not in the document]
Comment (scottbommarito) [reply src/NuGet.Services.Metadata.Catalog.Monitoring/Validation/Test/Registration/RegistrationDeprecationValidator.cs]: thanks a ton for this improvement, i was struggling with figuring out the best way and with your suggestion it's so much more readable now

File: src/NuGet.Services.Metadata.Catalog.Monitoring/Validation/Test/Registration/RegistrationDeprecationValidator.cs (hunk 5 above)
Comment (scottbommarito) [reply src/NuGet.Services.Metadata.Catalog.Monitoring/Validation/Test/Registration/RegistrationDeprecationValidator.cs]: I moved the alternate package verifications to their own function and then used the structure you suggested above

File: src/NuGet.Services.AzureSearch/SearchService/InvalidSearchRequestException.cs [hunk 9]
```diff
@@ -1,8 +1,4 @@
 ﻿using System;
-using System.Collections.Generic;
-using System.Linq;
-using System.Text;
-using System.Threading.Tasks;
```
Comment (agr) [inline src/NuGet.Services.AzureSearch/SearchService/InvalidSearchRequestException.cs]: File is missing copyright header.

File: src/NuGet.Services.Metadata.Catalog.Monitoring/Utility/ContainerBuilderExtensions.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (scottbommarito) [reply src/NuGet.Services.Metadata.Catalog.Monitoring/Utility/ContainerBuilderExtensions.cs]: My bad, resolved this accidentally.

Yes, `FeedType.HttpV2` is necessary. We are using client's APIs to fetch information to V2 and V3, and have created a dictionary of `FeedType` to client DI instance. So, `FeedType.HttpV3` maps to a client DI instance set up to download from V3, and `FeedType.HttpV2` maps to a client DI instance set up to query V2.

I don't think we actually use any of the client APIs for V2 anymore now that we query the gallery, so we _could_ replace all these DI dictionary shenanigans, but I'd rather not because it's not necessary. When I responded earlier I was thinking that I could replace the name `FeedType.HttpV2` to something like `FeedType.Database` or something to make things clearer, but I had forgotten that `FeedType` is defined in `NuGet.Protocol` and not in this project.

[the remark this answers is not in the document]
Comment (xavierdecoster) [reply src/NuGet.Services.Metadata.Catalog.Monitoring/Utility/ContainerBuilderExtensions.cs]: Gotcha! 👍
